### PR TITLE
Restores unavailable saltworks blueprints and recipes

### LINKED ISSUE #50623: Basecamp saltworks is missing its cooking recipes
### Describe the bug

The basecamp saltworks expansion has no available cooking recipes even when fully completed

### Steps To Reproduce

1. Mutate debug hammerspace
2. rapidly build base and saltworks (I tried evac shelter and field with same results)
3. no recipes

### Expected behavior

I expected there to be a recipe to craft salt from salt water. The JSONs for the saltworks and basecamp recipe_groups indicate that you should be able to craft salt and salt fish.

### Screenshots

_No response_

### Versions and configuration

- OS: Windows 10
- Game Version: 7a64f7e, the latest listed experimental release
- Graphics version: Tiles
- Ingame language: English
- Mods loaded: all default

### Additional context

Copying the recipe line for crafting "salt_from_salt_water" from the saltworks recipe group into the kitchen 1 recipe groups successfully added salt crafting to the game from the central building and canteen. ALSO, The sealab basecamp has the recipe group from the saltworks as it's supposed to.

### LINKED ISSUE #50625: Saltworks brewery building in basecamp can't be upgraded and has no recipes
### Describe the bug

Brewery upgrade can be built on basecamp but can't be upgraded or ordered to perform recipes

### Steps To Reproduce

1. Mutate debug hammerspace
2. rapidly build base and saltworks
3. build brewery building
4. no upgrades or recipes

### Expected behavior

The basecamp recipes json suggest and furniture of the basic brewery suggest it should be able to ferment and also there are missing upgrades in the json as well.

### Screenshots

_No response_

### Versions and configuration

OS: Windows 10
Game Version: 7a64f7e, the latest listed experimental release, slightly modified by me*
Graphics version: Tiles
Ingame language: English
Mods loaded: all default

### Additional context

Adding `{"id":"saltworks_recipes_2"}` to the `"blueprint_provides"` line for the variations of the basic brewery building seems to restore upgrades and recipes exactly as they should, so fix found I suppose barring and issues!

*In the course of trying to another bug https://github.com/CleverRaven/Cataclysm-DDA/issues/50623 the saltpan building now has the `{ "id": "kitchen" }` tag

### PR DESCRIPTION
**Pull request by Secret-Cisco**
#### Summary
Bugfixes "Restores unavailable saltworks blueprints and recipes"

#### Purpose of change
Fixes #50623 and Fixes #50625 and makes saltpan building blueprint tags more consistent.

#### Describe the solution
Adds { "id": "kitchen" } to saltpan building to allow it to use its cooking recipes properly and {"id":"saltworks_recipes_2"} to the brewery to allow access to its upgrades and crafting recipes (that at present do not work in a useful manner #50627)

Changed visible description of liquid brewery recipes to include WIP to warn players away as they don't yet work well. 

Also added { "id": "Salt_Pan" } to two material versions that did not have it so it was consistent across all saltpan buildings.

#### Describe alternatives you've considered
Maybe give the brewery recipes a different warning or delete them entirely. I don't know if the recipe group can be left empty and if the recipes get fixed in the future they would have to be re-added.

#### Testing
Made these edits to a downloaded copy and found they fixed the problem. Testing steps and results can be seen in the bug reports.

#### Additional context
I'm very much new to this so I hope I did it right.

### CODE REVIEW

**Comment by actual-nh** [comment]
If all you're changing is JSON, you don't need to recompile the game, BTW. To fix the PR summary, try (in place of `Category "Bugfixes"`) either `None` or `Bugfixes "Restores unavailable saltworks blueprints and recipes"`.

**Comment by Secret-Cisco** [comment]
Right, all I changed was the JSONs. I didn't think I really needed to compile the game, just that I dunno I couldn't test the whole repository as it is? I changed the summary, is that right now?

**Comment by actual-nh** [comment]
> Right, all I changed was the JSONs. I didn't think I really needed to compile the game, just that I dunno I couldn't test the whole repository as it is?

Well, you don't have the test program compiled, but you shouldn't need to.

> I changed the summary, is that right now?

Almost - just remove the "Category".

**Comment by Secret-Cisco** [comment]
"Category" removed. It says the JSON style check failed and I don't know what that mean but it seems suboptimal. Is that anything I need to worry about or can reasonably do to avoid that in the future?

**Comment by actual-nh** [comment]
> "Category" removed. It says the JSON style check failed and I don't know what that mean but it seems suboptimal. Is that anything I need to worry about or can reasonably do to avoid that in the future?

See previous reply, at the end. 🙂

**Comment by actual-nh** [comment]
The JSON style check is to make sure everything is consistent and there aren't apparent changes just because of spacing.

**Comment by Secret-Cisco** [comment]
OH, was that there the whole time... My bad, I can fix that up. Thanks for the help! I gotta ask though, is there any quicker/easier way to make the same identical change to 7 very similar files than doing it one at a time?

**Comment by actual-nh** [comment]
(No, it wasn't there the entire time - possibly too little time, sorry.) It would depend on the editor. If you can open multiple files and do a search-and-replace on all of them, that would work in some circumstances.

**Comment by Secret-Cisco** [comment]
Ah, I could do that with Notepad++ but I think I'd need to clone the repository to desktop github first? Waiting for that to download seems like more trouble than just doing on the website 7 times tbh.

**Comment by actual-nh** [comment]
Oh. One other thing. For this to show up as closing #50625, it has to have "Fixes" right before it - github's system doesn't understand "and".

**Comment by Secret-Cisco** [comment]
Okay, now it should now be properly linted and linked! Do I need to do anything to update the pull request after the new commits? It seems to have them already.

**Comment by actual-nh** [comment]
> Okay, now it should now be properly linted and linked! Do I need to do anything to update the pull request after the new commits? It seems to have them already.

The PR itself indeed has them already - it's going off of what's currently in your branch.

**Comment by Secret-Cisco** [comment]
Apparently there's supposed to be a space after `"id":` too 😅 Back to the formatting board.

**Comment by actual-nh** [comment]
If/when you do grab the entire repository, you can compile a formatting tool, actually.

**Comment by Maleclypse** [comment]
I think the solution here is to follow the `"suffix_id"` format.
```
  {
    "result": "offal_canned",
    "type": "recipe",
    "activity_level": "LIGHT_EXERCISE",
    "id_suffix": "jarred",
    "byproducts": [ [ "water", 10 ] ],
    "category": "CC_FOOD",
    "subcategory": "CSC_FOOD_MEAT",
    "skill_used": "cooking",
    "difficulty": 4,
    "time": "30 m",
    "batch_time_factors": [ 83, 5 ],
    "book_learn": [ [ "cookbook", 4 ], [ "manual_canning", 3 ] ],
    "container": "jar_glass_sealed",
    "result_mult": 2,
    "qualities": [ { "id": "CUT", "level": 1 }, { "id": "COOK", "level": 3 } ],
    "tools": [ [ [ "surface_heat", 100, "LIST" ] ], [ [ "pot_canning", -1 ] ] ],
    "components": [ [ [ "water", 11 ], [ "water_clean", 11 ] ], [ [ "jar_glass_sealed", 1 ] ], [ [ "meat_offal", 2, "LIST" ] ] ]
  },
  {
    "result": "offal_canned",
    "type": "recipe",
    "activity_level": "LIGHT_EXERCISE",
    "category": "CC_FOOD",
    "subcategory": "CSC_FOOD_MEAT",
    "skill_used": "cooking",
    "difficulty": 5,
    "time": "24 m",
    "batch_time_factors": [ 83, 5 ],
    "book_learn": [ [ "manual_sealing", 4 ], [ "atomic_survival", 4 ] ],
    "contained": true,
    "qualities": [
      { "id": "SAW_M", "level": 1 },
      { "id": "HAMMER", "level": 1 },
      { "id": "CUT", "level": 1 },
      { "id": "COOK", "level": 3 }
    ],
    "tools": [ [ [ "surface_heat", 10, "LIST" ] ], [ [ "can_sealer", -1 ] ] ],
    "components": [
      [ [ "canister_empty", 1 ], [ "can_medium", 1 ] ],
      [ [ "scrap", 1 ] ],
      [ [ "water", 1 ], [ "water_clean", 1 ] ],
      [ [ "meat_offal", 1, "LIST" ] ]
    ]
  },
```
Create new recipes that are `"brew_xxx_kegged"`  using either the `"container"` or `"contained": true` and follow the recipe exactly except they require a wooden or aluminum keg. Maybe that would fix them?  I'd try to get one working before I made changes to all the recipes.

**Comment by Maleclypse** [comment]
I think it would be acceptable to possibly merge this PR as is and create a new issue describing the problem with containing the liquid recipes along wiht my suggested solution if my solution is too much scope creep.

**Comment by Secret-Cisco** [comment]
I did a really quick and dirty test and hey would you look at that! 
image

**Comment by Secret-Cisco** [comment]
Guess it'll need new basecamp specific contained recipes then. All the rest of the NPC only recipes for that seem to be in recipe_companion.json so I suppose that's where these should go too?

Just gotta figure out the right container and such. Hmm can batches all be put in one container like this? Or maybe one batch of the special recipe should just be big enough to fill a decent sized container because it seems like you'd have to put musts in the fermenting vat yourself and dumping a hundred plastic bottles in there seems terrible.

**Comment by Secret-Cisco** [comment]
Adding `"contained": true` to the normal burdock wine must recipe and now NPC materializes a new glass bottle for every recipe like spawning it contained with the debug menu.

I also learned that you can only add must that you're carrying to the vat, so that definitely goes into the calculations on what container to use. Too small and you need a ton of containers and there's lots of juggling to get them into the vat, too big and you might not be able to lift it for use or find enough material to make so much in one go. Gallon jugs seem pretty common?

**Comment by Secret-Cisco** [comment]
Someone just made me remember that the kitchen also has a liquid recipe and I think that maybe a few more should be added like boiling water in the basic kitchen and maybe lamp oil production or other fuels like biodiesel? Containing the extant recipes and maybe adding water seems fine but I don't want to commit to adding the fuels too.

Funny thing is that I actually play the Bright Nights fork of DDA and all of this is because I wanted thought it prudent to test the new saltworks features before trying to port them!

**Comment by Secret-Cisco** [comment]
> If/when you do grab the entire repository, you can compile a formatting tool, actually. See https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/JSON_STYLE.md.

I'll be honest with you, I've read that page something like a dozen times and have yet to figure out how to make the formatter work but I did download Visual Studio and Visual Studio Code to mess around with.

**Comment by Secret-Cisco** [comment]
While I'm basically justspitballing ideas to add recipes to buildings, maybe the livestock expansion could make fertilizer and adobe brick. Maybe the garage could cook up biodiesel, it is car related.

**Comment by Maleclypse** [comment]
> Adding `"contained": true` to the normal burdock wine must recipe and now NPC materializes a new glass bottle for every recipe like spawning it contained with the debug menu.
> 
> I also learned that you can only add must that you're carrying to the vat, so that definitely goes into the calculations on what container to use. Too small and you need a ton of containers and there's lots of juggling to get them into the vat, too big and you might not be able to lift it for use or find enough material to make so much in one go. Gallon jugs seem pretty common?

I think the blueprint builds a wooden keg or two at the saltworks that you could use as the container by disassembling them and moving them to storage. But I’d have to check to be sure.

**Comment by Maleclypse** [comment]
> > If/when you do grab the entire repository, you can compile a formatting tool, actually. See https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/JSON_STYLE.md.
> 
> I'll be honest with you, I've read that page something like a dozen times and have yet to figure out how to make the formatter work but I did download Visual Studio and Visual Studio Code to mess around with.

If you want someone to help you in real time there are usually people on the discord channel #dev-help. Who understand how to set up an environment.

**Comment by Maleclypse** [comment]
> Someone just made me remember that the kitchen also has a liquid recipe and I think that maybe a few more should be added like boiling water in the basic kitchen and maybe lamp oil production or other fuels like biodiesel? Containing the extant recipes and maybe adding water seems fine but I don't want to commit to adding the fuels too.
> 
> Funny thing is that I actually play the Bright Nights fork of DDA and all of this is because I wanted thought it prudent to test the new saltworks features before trying to port them!

We appreciate your help on this!

**Comment by curstwist** [comment]
All great suggestions for recipe additions.  NPC camp recipes are under developed.  I keep hoping someone adds a egg collection zone as well.

All the `blueprint_provides` updates make sense and look good.

**Comment by Secret-Cisco** [comment]
> > Adding `"contained": true` to the normal burdock wine must recipe and now NPC materializes a new glass bottle for every recipe like spawning it contained with the debug menu.
> > I also learned that you can only add must that you're carrying to the vat, so that definitely goes into the calculations on what container to use. Too small and you need a ton of containers and there's lots of juggling to get them into the vat, too big and you might not be able to lift it for use or find enough material to make so much in one go. Gallon jugs seem pretty common?
> 
> I think the blueprint builds a wooden keg or two at the saltworks that you could use as the container by disassembling them and moving them to storage. But I’d have to check to be sure.

I checked and the keg dismantles into planks and such, so there's no equivalent container item it turns into.  I did try spawning in a wooden barrel and I learned 2 things:
1.  It actually takes surprisingly little materials to fill up a 100L (400 units) barrel with must/wort. It's about 19 batches (21 portions each for 399 total) of most of the recipes, which requires about 19 yeast, 57 water, a few hundred sugar/honey and whatever the base is.
2. You can lift a full barrel if you have a nearby crane/etc

So it's definitely possible doing in bulk if not necessarily available as such right away, but then neither is the brewery building.

**Comment by Secret-Cisco** [comment]
Oh, making musts/worts is super quick. It's usually only about an hour and a half to make a thousand units by hand so the delegated crafting should definitely be in big batches, anything smaller and you might as well just do it yourself.

I'm thinking make the recipes require and be stored in a wooden barrel, use the time and ingredients it'd to to make 19 normal batches (for applicable recipes, some differ) and round it up to an even 400.

Wooden barrel because I feel it matches the brewery style, and it's craftable. It only requires planks, scrap, an anvil, and fabrication lvl 3. It takes 6 hours to make though, but I was thinking it could be very easily added to the **fabrication workshop** for manufacture. The other craftable containers with similar capacity (i.e. the steel keg, drums, and plastic barrel) all require fabrication 8, which seems a bit much to ask for.

**Comment by Secret-Cisco** [comment]
First prototype looking good! It doesn't seem that I can make it a round number but this is good enough.
image
image

File: data/json/recipes/basecamps/recipe_groups.json [hunk 1]
```diff
@@ -201,23 +201,23 @@
     "id": "saltworks_recipes_2",
     "building_type": "COOK",
     "recipes": [
-      { "id": "brew_fruit_wine", "description": " Brew: Fruit Wine Must" },
-      { "id": "brew_mead", "description": " Brew: Spiced Wine Must" },
-      { "id": "brew_dandelion_wine", "description": " Brew: Dandelion Wine Must" },
-      { "id": "brew_burdock_wine", "description": " Brew: Burdock Wine Must" },
-      { "id": "brew_pine_wine", "description": " Brew: Pine Wine Must" },
-      { "id": "brew_hb_beer", "description": " Brew: Beer" }
+      { "id": "brew_fruit_wine", "description": " WIP Brew: Fruit Wine Must" },
```
Comment (Maleclypse) [inline data/json/recipes/basecamps/recipe_groups.json:204]: Do these still need to be renamed as WIP or does the solution work as expected?   It looks good from what I saw of it in your screenshots.

**Review by Maleclypse — COMMENTED** [review]
I was going to ask if this was ready and looking over I realize that you haven't pushed your local changes using the wooden barrels yet.   :) I'm super excited with how this is looking from your screenshots.

**Comment by Secret-Cisco** [comment]
> I was going to ask if this was ready and looking over I realize that you haven't pushed your local changes using the wooden barrels yet. :) I'm super excited with how this is looking from your screenshots.

Yeah I hit a big snag on that. I wrote out all the contained versions + added more for moonshine and the hard liquor worts, and tested them out fine except that I got the portions wrong and found out the volume of worts/musts/etc portions are much smaller in volume than water.

So the recipes work technically but the 400 portions I expected to fill the barrel are only ~15 L in a 100 L barrel lmao. So now I'm gonna rewrite them all for a different, as yet undecided container*, so they actually fit is something not entirely absurd.


* Containers I'm considering
1. A clay hydria is exactly 15 L so I wouldn't have to rewrite anything else but the container, but it doesn't feel right and can't really be found in the world.
2. A steel jerry can is 20 L so it'd only be 25% empty instead of 85% without a rewrite, and with rewriting would still be reasonable size.
3. A gallon jug is much smaller and would require completely rewriting the recipes, I don't like their small size because you'd need a lot of them but they are fairly common and craftable.
4. Something else, I dunno.

**In summary**: the recipes technically work but are very silly due to a simple mistake so I took a break to think about how to fix it and to actually play the game.

**Comment by Secret-Cisco** [comment]
Beer wort is actually the same volume as water though and crafts in multiples of 10 so it fits perfectly into a wood barrel as is.

**Comment by Secret-Cisco** [comment]
I've given up (for now at least) on making the brewing recipes work in a sane way, the inconsistency between liquids/recipes and pocket limitations are making me insane and I frankly no longer have the drive to want to work on this problem. I left a note in the basecamp recipe groups JSON for any future intrepid contributor.

So I'm just contributing the original bugfixes & such, hopefully they don't conflict with the current state of the repository.

**Review by Maleclypse — APPROVED** [review]
